src/java.base/share/classes/jdk/internal/vm/annotation/Stable.java

```diff
@@ -28,56 +28,191 @@
- * <p>
- * Since all fields begin with a default value of null for references
- * (resp., zero for primitives), it follows that this annotation indicates
- * that the first non-null (resp., non-zero) value stored in the field
- * will never be changed.
- * <p>
- * If the field is not of an array type, there are no array elements,
- * then the value indicated as stable is simply the value of the field.
- * If the dynamic type of the field value is an array but the static type
- * is not, the components of the array are <em>not</em> regarded as stable.
- * <p>
- * If the field is an array type, then both the field value and
- * all the components of the field value (if the field value is non-null)
- * are indicated to be stable.
- * If the field type is an array type with rank {@code N > 1},
- * then each component of the field value (if the field value is non-null),
- * is regarded as a stable array of rank {@code N-1}.
+ * A field may be annotated as "stable" to indicate that it is a
+ * <em>stable variable</em>, expected to change value at most once.
+ * The first value stored into the field (assuming it is not a
+ * duplicate of the the field's default initial value), allows the VM
+ * to assume that no more significant changes will occur.  This in
+ * turn enables the VM to optimize the stable variable, treating uses
+ * of it as constant values.  This behavior is a useful building block
+ * for lazy evaluation or memoization of results.  In rare and subtle
+ * use cases, stable fields may also assume on multiple values over
+ * time, with effects to be described below.
+ * <p>
+ * For example, declaring two stable fields of type {@code int} and
+ * {@code String} creates a pair of stable variables, initialized to
+ * zero and the null reference (respectively).  Storing a non-zero
+ * integer to the first field, or a non-null string to the second,
+ * will enable the VM to expect that the stored value is now the
+ * permanent value of the field, going forward.  This condition may be
+ * used by the VM compiler to improve code quality more aggressively.
+ * <p>
+ * Since all heap variables begin with a default null value for
+ * references (resp., zero for primitives), there is an ambiguity when
+ * the VM discovers a stable variable holding a null or primitive zero
+ * value.  Does the user intend the VM to constant fold that
+ * (uninteresting) value?  Or is the user waiting until later to
+ * assign another value to the variable?  The VM does not
+ * systematically record stores of a default null (or primitive zero),
+ * so there is no way for the VM to decide if a default field value is
+ * an undisturbed initial default value, or has been overwritten with
+ * an intentionally stored null (or primitive zero).  This is why the
+ * programmer should store non-default values into stable variables,
+ * if the consequent optimization is desired.
+ * <p>
+ * As a special case, if a stable field is declared as an array type
```

Comment:
If `array` is a stable array field, is `array.length` treated as an independent stable variable?

Reply:
If `array` is a stable field, wouldn't `array.length` just be a (potentially compile time) constant?